File: Realize/6Week/PracticeGraph2.cpp
// Practice1
// 인접 리스트를 이용한 그래프 구현

#include <iostream>
#include <vector>
using namespace std;

struct Node {
    int id;
    Node* next;
    Node(int id, Node* next = nullptr) : id(id), next(next) {}
};

class MyGraphList {
    vector<char> vertices;
    vector<Node*> adjList;
    int elemCnt;
    int size;

public:
    MyGraphList(int size) : size(size), elemCnt(0) {
        vertices.resize(size);
        adjList.resize(size, nullptr);
    }

    bool isFull() {
        return elemCnt == size;
    }

    void addVertex(char data) {
        if (isFull()) {
            cout << "Graph is Full!!" << endl;
            return;
        }
        vertices[elemCnt++] = data;
    }

    void addEdge(int x, int y) {
        adjList[x] = new Node(y, adjList[x]);
        adjList[y] = new Node(x, adjList[y]);
    }

    void addDirectedEdge(int x, int y) {
        adjList[x] = new Node(y, adjList[x]);
    }

    void printAdjacentList() {
        for (int i = 0; i < elemCnt; i++) {
            cout << vertices[i] << ": ";
            Node* cur = adjList[i];
            while (cur != nullptr) {
                cout << vertices[cur->id] << " ";
                cur = cur->next;
            }
            cout << endl;
        }
    }
};

int main() {
    MyGraphList graph(4);

    graph.addVertex('A');
    graph.addVertex('B');
    graph.addVertex('C');
    graph.addVertex('D');

    graph.addEdge(0, 1);
    graph.addEdge(0, 2);
    graph.addEdge(1, 2);
    graph.addEdge(1, 3);
    graph.addEdge(2, 3);
    graph.printAdjacentList();
}

File: Realize/6Week/PracticeGraph2.py
# Practice1
# 인접 리스트를 이용한 그래프 구현
class Node:
    def __init__(self, id, next=None):
        self.id = id
        self.next = next

class MyGraphList:
    def __init__(self, size):
        self.vertices = [None] * size
        self.adjList = [None] * size
        self.elemCnt = 0
        self.size = size

    def is_full(self):
        return self.elemCnt == self.size

    def add_vertex(self, data):
        if self.is_full():
            print("Graph is Full!!")
            return
        self.vertices[self.elemCnt] = data
        self.elemCnt += 1

    def add_edge(self, x, y):
        self.adjList[x] = Node(y, self.adjList[x])
        self.adjList[y] = Node(x, self.adjList[y])

    def add_directed_edge(self, x, y):
        self.adjList[x] = Node(y, self.adjList[x])

    def print_adjacent_list(self):
        for i in range(self.elemCnt):
            print(f"{self.vertices[i]}: ", end="")
            cur = self.adjList[i]
            while cur:
                print(self.vertices[cur.id], end=" ")
                cur = cur.next
            print()

# Test code
graph = MyGraphList(4)
graph.add_vertex('A')
graph.add_vertex('B')
graph.add_vertex('C')
graph.add_vertex('D')

graph.add_edge(0, 1)
graph.add_edge(0, 2)
graph.add_edge(1, 2)
graph.add_edge(1, 3)
graph.add_edge(2, 3)
graph.print_adjacent_list()
